contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,238 @@
+/// and returns a DatabaseConfig struct.
+///
+/// Example usage:
+/// ```rust,ignore
+/// let config = database_config("my_database", rocket.config());
+/// ```
+pub fn database_config<'a>(name: &str, from: &'a config::Config) -> Option<DatabaseConfig<'a>> {
+    let database_configs = match from.get_slice("databases") {
+        Ok(configs) => configs,
+        Err(_) => {
+            error_!("Rocket.toml does not contain any database configurations when attempting to fetch database: {}", name);
+            return None;
+        }
+    };
+
+    let connection_config = match database_configs
+        .iter()
+        .filter(|e| {
+            match e.as_table().map(|table| {
+                table.get("name").map(|v| v.as_str()) == Some(Some(name))
+            }) {
+                Some(true) => true,
+                _ => false,
+            }
+        })
+        .nth(0) {
+        Some(config) => config,
+        None => {
+            error_!("Could not find any database configurations named: {}", name);
+            warn_!("Available configurations: {}", available_database_configs(from).join(", "));
+            return None;
+        }
+    };
+
+    let url = match connection_config.get("url") {
+        Some(url) => {
+            match url.as_str() {
+                Some(url) => url,
+                None => {
+                    error_!("Could not convert the URL into a valid string");
+                    return None;
+                }
+            }
+        }
+        None => {
+            error_!("Could not fetch the URL as part of the connection config");
+            return None;
+        }
+    };
+
+    let pool_size = match connection_config.get("pool_size") {
+        Some(pool_size) => {
+            pool_size.as_integer().unwrap_or(from.workers as i64) as u32
+        }
+        None => 10,
+    };
+
+    let extras = connection_config.as_table();
+
+    Some(DatabaseConfig {
+        url,
+        pool_size,
+        extras,
+    })
+}
+
+/// Returns an iterator containing the names of all the possible database connections.
+/// The primary use within `rocket_contrib` for this method is in relation to calling out
+/// valid database connections that are specified in the `Rocket.toml` file.
+pub fn available_database_configs<'a>(config: &'a config::Config) -> Vec<&'a str> {
+    let db_config = match config.get_slice("databases") {
+        Ok(config) => config,
+        Err(_) => return vec![],
+    };
+
+    db_config
+        .iter()
+        .map(|config| {
+            config.as_table().and_then(|table| table["name"].as_str())
+        })
+        .filter_map(|e| e)
+        .collect::<Vec<&'a str>>()
+}
+
+/// This trait can be implemented on Connection types that can be stored in a connection pool.
+/// Implementing this trait on a type allows Rocket to generate the necessary code to manage the
+/// pool internally and inject it into request handlers.
+pub trait Poolable: Send + Sized + 'static {
```

Comment:
I'm not super excited about the dependency on r2d2 and `ManageConnection`, but defining something similar to r2d2's existing trait hierarchy might be even more headache. It is easy enough to adapt r2d2 for arbitrary types like `Arc<T>` if it's really necessary for some use case, after all.

Reply:
I'm okay with pivoting direction here, but to me, the original PR felt like r2d2 was a key implementation detail. I would want us to have a larger conversation about this decision, but I'm not opposed to removing `ManageConnection` as a key dependency. I just want to know that this is the correct direction to take this PR.

Reply:
I thought about it more and I think r2d2 is a good thing to build on, given the prevalence and flexibility of r2d2. Any use cases that are truly incompatible with r2d2 can implement pool and connection types manually as before.

Reply:
They just won't necessarily be able to leverage our code generation facilities.

This is also just a starting place. Once it's polished enough for inclusion in `rocket_contrib`, we absolutely can iterate on it.